### PR TITLE
loading spinner

### LINKED ISSUE #208: Show loading spinner while frame differences are computed (Library 3)
## User story
1. As a user
2. I want to see a loading spinner while the frame differences are computed
3. So that I know that the software is still doing something

## Acceptance criteria
* after clicking on the button: "Compute and Display Diferences" a loading spinner should be shown as long as the computation takes
* below the loading spinner there should be a 'X'
* when clicking the 'X' the computation gets canceled
* while the loading spinner is active the background should be inactive and darker
* the loading spinner should look similar to this:
Image
* when the computation is finished and the program jumps to the video player screen the loading spinner should disappear

## Definition of done (DoD)
* Create feature branch
* Implement the functionality
* Opened pull request and assigned reviewer
* If necessary: required changes are implemented
* Delete feature branch
* If necessary: Component Test has been written
* Code review has been completed and code has been merged

## DoD general criteria
* Feature has been fully implemented
* Feature has been merged into the mainline
* All acceptance criteria were met
* Product owner approved features
* All tests are passing
* Developers agreed to release

### PR DESCRIPTION
**Pull request by AlperK61**
closes #208

### CODE REVIEW

**Review by a-miscellaneous — CHANGES_REQUESTED** [review]

**Comment by a-miscellaneous** [comment]
generating differences is sadly not cancelable, job.cancel does not work.
parentJob lauch generation and then parentjob.cancel also does not work

**Comment by akriese** [comment]
What do we do with this then? @a-miscellaneous @AlperK61

**Comment by a-miscellaneous** [comment]
> What do we do with this then? @a-miscellaneous @AlperK61

next sprint, conflicts left

**Comment by akriese** [comment]
We have to make the difference generation cooperative for cancelling. Then, `job.cancel()` will work. I'll look into that now

**Comment by akriese** [comment]
After doing my research, it is pretty clear, that we can only cancel the job by injecting some kind of suspend function into the `Algorithm.run()` or some check if the job is still wanted. That would mean needing to refactor lib2 quite a bit so that it is cancel-aware... Unfortunately, the builtin `Threat.yield()` does not work as `kotlinx.coroutines.yield` and thus, lib2 must include kotlinx somehow.
Or... We redefine the algorithm to run in steps, like a ML training. Then, we could check if the job was cancelled between each step.
Some links I checked:
https://stackoverflow.com/questions/69097765/how-to-stop-or-cancel-a-kotlin-coroutine-stop-currently-running-code-block-inst
https://stackoverflow.com/questions/74065990/properly-cancelling-kotlin-coroutine-job

**Comment by akriese** [comment]
Rebased onto main and added singleton to cancel running algorithm instance.

File: GUI/src/main/kotlin/ui/components/selectVideoScreen/LoadingDialog.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import androidx.compose.material3.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.unit.dp
+
+@OptIn(ExperimentalMaterialApi::class)
+@Composable
+fun LoadingDialog(onCancel: () -> Unit = {}) {
+    AlertDialog(
```
Comment (zino212) [inline GUI/src/main/kotlin/ui/components/selectVideoScreen/LoadingDialog.kt:20]: The other AlertDialogs we use are now styled like this:

image

You can either change that here or is will be done in the design branch after rebasing on this one. If requested I can also change this here.

**Review by zino212 — CHANGES_REQUESTED** [review]
LGTM and seems to work nicely. I only request using the Material 3 Colorscheme and Fonts for the Dialog

**Review by zino212 — APPROVED** [review]

**Pull request merged**